internal/driver/testhelpers.go

```diff
@@ -0,0 +1,63 @@
+package driver
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ory/keto/internal/driver/config"
+)
+
+type GetAddr = func(testing.TB, string) (host string, port string, fullAddr string)
+
+func UseDynamicPorts(ctx context.Context, t testing.TB, r Registry) GetAddr {
```

Comment:
No not really. The test has to do these steps in this order:

1. create the registry
2. call `UseDynamicPorts` to set the config values
3. start the server(s); potentially only some and not all
4. use the closure to retrieve the actual address the server is listening on

What we do with this change is let the OS choose a free port when the server starts listening (using port `0`). We can only get the actual port number after the server was started, therefore this step comes later.
The problem with `freeport` is that their approach is to open a socket on port `0`, get the actual port number, and then close it again. This results in racy behavior where a parallel test gets the same port assigned, because the socket is closed and then reopened.